src/frontend/src/optimizer/property/func_dep.rs

```diff
@@ -0,0 +1,373 @@
+    /// considered as equal. So for following table, A --> B is not valid.
+    /// **NOT** allowed.
+    /// ```text
+    ///   A   | B
+    /// ------|---
+    ///  NULL | 1
+    ///  NULL | 2
+    /// ```
+    ///
+    /// Currently we only have strict dependencies, but we may also have lax dependencies in the
+    /// future.
+    strict: Vec<FunctionalDependency>,
+}
+
+impl fmt::Display for FunctionalDependencySet {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("{")?;
+        for fd in self.strict.iter().with_position() {
+            match fd {
+                itertools::Position::First(fd) | itertools::Position::Middle(fd) => {
+                    f.write_fmt(format_args!("{}, ", fd))?;
+                }
+                itertools::Position::Last(fd) | itertools::Position::Only(fd) => {
+                    f.write_fmt(format_args!("{}", fd))?;
+                }
+            }
+        }
+        f.write_str("}")
+    }
+}
+
+impl FunctionalDependencySet {
+    /// Create a empty [`FunctionalDependencySet`]
+    pub fn new() -> Self {
+        Self { strict: Vec::new() }
+    }
+
+    /// Create a [`FunctionalDependencySet`] with the indices of a key.
+    ///
+    /// The **combination** of these columns can determine all other columns.
+    ///
+    /// Note that if `key_indices` empty, an empty fd set will be returned.
+    ///
+    /// # Examples
+    /// ```rust
+    /// # use risingwave_frontend::optimizer::property::{FunctionalDependencySet, FunctionalDependency};
+    /// # use itertools::Itertools;
+    /// let mut fd = FunctionalDependencySet::with_key(4, &[1]);
+    /// let fd_inner = fd.into_dependencies();
+    ///
+    /// assert_eq!(fd_inner.len(), 1);
+    ///
+    /// let FunctionalDependency { from, to } = &fd_inner[0];
+    /// // 1 --> 0, 2, 3
+    /// assert_eq!(from.ones().collect_vec(), &[1]);
+    /// assert_eq!(to.ones().collect_vec(), &[0, 2, 3]);
+    /// ```
+    pub fn with_key(column_cnt: usize, key_indices: &[usize]) -> Self {
+        let mut tmp = Self::new();
+        if !key_indices.is_empty() {
+            tmp.add_key_column(column_cnt, key_indices);
+        }
+        tmp
+    }
+
+    /// Create a [`FunctionalDependencySet`] with a dependency [`Vec`]
+    pub fn with_dependencies(strict_dependencies: Vec<FunctionalDependency>) -> Self {
+        Self {
+            strict: strict_dependencies,
+        }
+    }
+
+    pub fn as_dependencies_mut(&mut self) -> &mut Vec<FunctionalDependency> {
+        &mut self.strict
+    }
+
+    pub fn as_dependencies(&self) -> &Vec<FunctionalDependency> {
+        &self.strict
+    }
+
+    pub fn into_dependencies(self) -> Vec<FunctionalDependency> {
+        self.strict
+    }
+
+    /// Add a functional dependency to a [`FunctionalDependencySet`].
+    pub fn add_functional_dependency(&mut self, fd: FunctionalDependency) {
```

Comment:
:thinking: I'm also wondering about the API design here. Maybe we should only keep `add_functional_dependency` in `FDSet`. So user will write `some_fd_set.add_functional_dependency(FunctionalDependency::with_indices(&[0, 1], &[2, 3]))`. But this might be too verbose... 

cc @st1page

Reply:
Alternatively, make FD completely private, and only expose FDset?

Reply:
> User will always work with FDset instead of single FDs

I think this is possible and will only affect unit testing for now. But I'm worried about whether this will affect future operators. i.e. maybe the fd inference logic for some operators requires fine control at fd level.